File: command-center-tauri/src/utils/json-pointer.ts
export type PointerResult = {
  value: unknown;
  exists: boolean;
  error?: string;
};

/// Resolve a JSON pointer (RFC 6901) or dotted/bracketed path against a value.
/// Accepts "/foo/bar", "foo.bar", "foo[0].bar". Returns { exists: false } when
/// any segment is missing; never throws.
export function evaluatePointer(input: unknown, pointer: string): PointerResult {
  const trimmed = pointer.trim();
  if (!trimmed) return { value: input, exists: true };

  const segments = parseSegments(trimmed);
  let current: any = input;
  for (const seg of segments) {
    if (current == null) return { value: undefined, exists: false };
    if (typeof current !== "object") return { value: undefined, exists: false };
    if (!(seg in current)) return { value: undefined, exists: false };
    current = current[seg];
  }
  return { value: current, exists: true };
}

function parseSegments(pointer: string): string[] {
  if (pointer.startsWith("/")) {
    return pointer
      .slice(1)
      .split("/")
      .map((s) => s.replace(/~1/g, "/").replace(/~0/g, "~"));
  }
  const segments: string[] = [];
  let buf = "";
  for (let i = 0; i < pointer.length; i++) {
    const ch = pointer[i];
    if (ch === ".") {
      if (buf) segments.push(buf);
      buf = "";
    } else if (ch === "[") {
      if (buf) segments.push(buf);
      buf = "";
      const end = pointer.indexOf("]", i);
      if (end < 0) break;
      const inner = pointer.slice(i + 1, end);
      segments.push(inner.replace(/^['"]|['"]$/g, ""));
      i = end;
    } else {
      buf += ch;
    }
  }
  if (buf) segments.push(buf);
  return segments;
}

File: runinator-command-center-flutter/lib/core/utils/json_pointer.dart
// port of core/utils/json-pointer.ts.

class PointerResult {
  const PointerResult({required this.value, required this.exists, this.error});

  final Object? value;
  final bool exists;
  final String? error;
}

/// resolve a JSON pointer (RFC 6901) or dotted/bracketed path against a value.
/// accepts "/foo/bar", "foo.bar", "foo[0].bar". returns exists: false when
/// any segment is missing; never throws.
PointerResult evaluatePointer(Object? input, String pointer) {
  final trimmed = pointer.trim();

  if (trimmed.isEmpty) {
    return PointerResult(value: input, exists: true);
  }

  final segments = _parseSegments(trimmed);
  Object? current = input;

  for (final seg in segments) {
    if (current == null) {
      return const PointerResult(value: null, exists: false);
    }

    if (current is List) {
      final index = int.tryParse(seg);
      if (index == null || index < 0 || index >= current.length) {
        return const PointerResult(value: null, exists: false);
      }
      current = current[index];
      continue;
    }

    if (current is! Map) {
      return const PointerResult(value: null, exists: false);
    }

    if (!current.containsKey(seg)) {
      return const PointerResult(value: null, exists: false);
    }

    current = current[seg];
  }

  return PointerResult(value: current, exists: true);
}

List<String> _parseSegments(String pointer) {
  if (pointer.startsWith('/')) {
    return pointer.substring(1).split('/').map((s) => s.replaceAll('~1', '/').replaceAll('~0', '~')).toList();
  }

  final segments = <String>[];
  var buf = '';

  var i = 0;
  while (i < pointer.length) {
    final ch = pointer[i];

    if (ch == '.') {
      if (buf.isNotEmpty) {
        segments.add(buf);
      }
      buf = '';
    } else if (ch == '[') {
      if (buf.isNotEmpty) {
        segments.add(buf);
      }
      buf = '';
      final end = pointer.indexOf(']', i);

      if (end < 0) {
        break;
      }

      final inner = pointer.substring(i + 1, end);
      segments.add(inner.replaceFirst(RegExp(r"""^['"]"""), '').replaceFirst(RegExp(r"""['"]$"""), ''));
      i = end;
    } else {
      buf += ch;
    }

    i++;
  }

  if (buf.isNotEmpty) {
    segments.add(buf);
  }

  return segments;
}
